File: 0001-0500/0300-最长递增子序列/java-0300/src/Solution2.java
public class Solution2 {
    // 贪心 + 二分
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        if (n == 0) {
            return 0;
        }

        // dp[i]表示长度为i的最长递增子序列的末尾元素最小值
        int[] dp = new int[n + 1];
        int len = 1;
        dp[len] = nums[0];

        for (int i = 1; i < n; i++) {
            if (nums[i] > dp[len]) {
                len++;
                dp[len] = nums[i];
            } else {
                // dp数组中[0, len]区间小于nums[i]的最大值索引
                int pos = lower(dp, 0, len, nums[i]);

                dp[pos + 1] = nums[i];
            }
        }
        return len;
    }

    private int lower(int[] nums, int l, int r, int target) {
        while (l < r) {
            int mid = l + (r - l + 1 >> 1);
            if (nums[mid] < target) {
                l = mid;
            } else {
                r = mid - 1;
            }
        }
        return l;
    }

    public static void main(String[] args) {
        int[] nums = {10,9,2,5,3,7,101,18};
        System.out.print(new Solution1().lengthOfLIS(nums));
    }
}

File: 剑指Offer/053Ⅰ-在排序数组中查找数字I/js-053Ⅰ/Solution2.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let lowerIdx = lower(nums, target);
    let upperIdx = upper(nums, target);

    return upperIdx - lowerIdx - 1;
};

// 查询小于target的最大值的索引
function lower(nums, target) {
    let l = -1;
    let r = nums.length - 1;

    while (l < r) {
        let mid = l + ((r - l + 1) >> 1);

        if (nums[mid] < target) {
            l = mid;
        } else {
            r = mid - 1;
        }
    }
    return l;
}

// 查询大于target的最小值的索引
function upper(nums, target) {
    let l = 0;
    let r = nums.length;

    while (l < r) {
        let mid = l + (r - l >> 1);

        if (nums[mid] > target) {
            r = mid;
        } else {
            l = mid + 1;
        }
    }
    return l;
}